api/lock.go

```diff
@@ -165,25 +154,36 @@ WAIT:
 	if pair != nil && pair.Flags != LockFlagValue {
 		return nil, ErrLockConflict
 	}
-	if pair != nil && pair.Session != "" {
+	if pair != nil && pair.Session != "" && pair.Session != l.lockSession {
 		qOpts.WaitIndex = meta.LastIndex
 		goto WAIT
 	}
 
-	// Try to acquire the lock
-	lockEnt := l.lockEntry(l.lockSession)
-	locked, _, err := kv.Acquire(lockEnt, nil)
-	if err != nil {
-		return nil, fmt.Errorf("failed to acquire lock: %v", err)
-	}
+	if pair == nil || pair.Session == "" {
+		if l.lockSession != "" {
```

Comment:
Just to further clarify this, let's say entity A acquires the lock. If it then loses the lock (and monitorLock notifies it of such), it will try to reacquire the lock by calling .Lock() again. Since monitorLock does not reset the l.lockSession, we still have the same lockSession and will try to reclaim the lock with that session. The server responds with an invalid session error in this case. I think there's an argument for doing all of the session work outside of the api (and that goes back to the previous comment about using a session ID from opts) but if we're already calling createSession internally, I think we should be recreating a valid session for the user in this case.